project_euler/problem_014/sol2.py

```diff
@@ -27,25 +27,27 @@
 """
 from __future__ import annotations
 
+COLLATZ_SEQUENCE_LENGTHS = {1: 1}
+
 
 def collatz_sequence_length(n: int) -> int:
     """Returns the Collatz sequence length for n."""
-    sequence_length = 1
-    while n != 1:
-        if n % 2 == 0:
-            n //= 2
-        else:
-            n = 3 * n + 1
-        sequence_length += 1
+    if n in COLLATZ_SEQUENCE_LENGTHS:
+        return COLLATZ_SEQUENCE_LENGTHS[n]
+    if n % 2 == 0:
+        next_n = n // 2
+    else:
+        next_n = 3 * n + 1
+    sequence_length = collatz_sequence_length(next_n) + 1
+    COLLATZ_SEQUENCE_LENGTHS[n] = sequence_length
```

Comment:
Unnecessary variable assignment.
This can be shortened to one line.

Reply:
But then we should return `COLLATZ_SEQUENCE_LENGTHS[n]` instead of `sequence_length`. I don't know Python perfectly - whether there will be an extra search in the dictionary or not?

Reply:
No, you're right. I missed line 43.
Disregard my previous comment.

I originally meant, that assigning the result to `sequence_length`  and immediately after assigning it back to `COLLATZ_SEQUENCE_LENGTH`, __without any further usage__ is unnecessary.

However, you're returning `sequence_length` as the result. Saving another lookup - in short, keep it this way. I was wrong.

Sorry for my ignorance.

Sincerely,
IdoErel